Hive Meta Store impersonation access

BlueStalker opened the pull request:
This PR enabled the option of impersonation access of the Hive Metastore impersonation.
It also includes the changes to enable multiple HMS instances load-balancing and reporting the stats for each individual
HMS instance.

Test plan - (Please fill in how you tested your changes)
Hands-on testing when enabling the impersonation
Plan to add more integration testing cases during turning on the config

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

```
== RELEASE NOTES ==

General Changes
N/A

Hive Changes
* Hive Metastore Impersonation Access, add configuration property ``hive.metastore-impersonation-enabled`` to turn it on
* Load balancing of multiple Hive Metastore instances, add configuration property ``hive.metastore.load-balancing-enabled`` to turn it on
```

BlueStalker commented:
ping @arhimondr Sorry that it has been a while since last time I worked on https://github.com/prestodb/presto/pull/14464, now that PR is staled since it has been a long time, I reopened it here. Please take another look, Thanks

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -125,43 +128,45 @@ public SemiTransactionalHiveMetastore(
         Action<TableAndMore> tableAction = tableActions.get(new SchemaTableName(databaseName, tableName));
         if (tableAction == null) {
-            return delegate.getTable(databaseName, tableName);
+            return metastoreAuthentication.doAs(identity.getUser(), () -> delegate.getTable(databaseName, tableName));
```
This is still too high level API to wrap in the `doAs` statement. As explained in the comment [here](https://github.com/prestodb/presto/pull/14464#issuecomment-642826246) different metastore implementations (Hive / Glue / Inmemory / Thrift / File / etc.) have completely different authentication methods. The `doAs` mechanism is HiveMetastore specific (even Hadoop client specific). Using it at this level breaks the encapsulation and abstraction.

Additionally, historically in Presto we are trying to minimize the usage of the thread local contexts, as they tend to be very fragile. Think if there's a metastore implementation that does method invokation in a separate thread pool for efficiency (e.g.: multiple calls in parallel). The `doAs` based implementation will be broken for such a use case.

Thus I strongly recommend you passing the identity all the way down to the lower level interface and wrapping the invokations with `doAs `only at the `ThriftHiveMetastore.class` level

BlueStalker replied:
Ping again, the doAs has been removed, and the new implementation is based on passing the MetastoreContext object explicitly in the interface.

aweisberg commented:
Sorry for the delay, the docker images were all deleted yesterday and I was trying to resolve that. I'll give reviewing this a shot. I don't have as much context as Andrii and people are pretty busy right now.

Is any of this a cherrypick from Trino?

aweisberg commented:
Since the product tests aren't working we can't properly test this right now. Some of Andrii's feedback from the last PR hasn't been incorporated like baking stuff into the docker images and default value in MetastoreClientConfig.

BlueStalker commented:
> Since the product tests aren't working we can't properly test this right now. Some of Andrii's feedback from the last PR hasn't been incorporated like baking stuff into the docker images and default value in MetastoreClientConfig.

Hey @aweisberg Thanks for reviewing, basically, this is the PR that originally from https://github.com/prestodb/presto/pull/14464, and this is not exactly the trino PR of https://github.com/trinodb/trino/pull/1441/files, 

In this PR, it does the HMS impersonation access like above trino PR did, as well ass adding metrics for each individual HMS, while the major concern from @arhimondr is the Thread Local Storage, I think I already addressed in the latest diff.

For the Docker Image problem, I am wondering what's the plan here, are we going to fix the docker image repo and upgrade the memory setting (basically bump up Xmx) in the image?

aweisberg commented:
Right now we can't build the docker images because Cloudera put everything behind a paywall. We are trying to address that, but it might take a bit.

I am asking about Trino because if there is code from Trino then we need co-author tags to reflect that.

aweisberg commented:
I'm sorry for the delay reviewing this. Both @arhimondr  and myself have been pretty swamped with things that have deadlines and this is a non-trivial review. Especially for me as I don't have much context in this area so I would have to build it.

BlueStalker commented:
> I'm sorry for the delay reviewing this. Both @arhimondr and myself have been pretty swamped with things that have deadlines and this is a non-trivial review. Especially for me as I don't have much context in this area so I would have to build it.

That's OK. this diff is NOT ONLY containing the changes for HMS impersonation but also including the metrics break-down by multiple-HMS since it also enables the RR multiple HMS access here.  It is a bit confusing for me how memory problems have been triggered by this diff. And when I tried to test it locally, as we discussed in the slack, even in loading the data, the yarn container has exited because of memeory issue, and I think the CI here has already been passed that step, it is not clear what's difference between local env and CI env which I feel like it is just a docker compose env..)

BlueStalker commented:
Ping @aweisberg @arhimondr again, this seemed to take quite a long time, what's our plan here.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java:
```diff
@@ -108,20 +111,22 @@ public MetastoreHiveStatisticsProvider(SemiTransactionalHiveMetastore metastore)
     }
 
-    private static Map<String, PartitionStatistics> getPartitionsStatistics(SemiTransactionalHiveMetastore metastore, SchemaTableName table, List<HivePartition> hivePartitions)
+    private static Map<String, PartitionStatistics> getPartitionsStatistics(String hmsImpersonationDefaultUser, SemiTransactionalHiveMetastore metastore, SchemaTableName table, List<HivePartition> hivePartitions)
```
Let's refactor to pass the identity from the ConnectorSession

arhimondr commented on presto-hive/pom.xml:
```diff
@@ -216,6 +216,11 @@
         </dependency>
 
+        <dependency>
+            <groupId>commons-codec</groupId>
```
We are trying to avoid using apache libraries. Why is this library needed? Can you achieve the same by using the codecs from Guava?

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java:
```diff
@@ -252,47 +253,47 @@ protected RoleGrant computeNext()
     }
 
-    public static boolean isRoleApplicable(SemiTransactionalHiveMetastore metastore, PrestoPrincipal principal, String role)
+    public static boolean isRoleApplicable(SemiTransactionalHiveMetastore metastore, ConnectorIdentity identity, PrestoPrincipal principal, String role)
```
nit: I would recommend passing the `MetastoreContext` directly to the methods in this class

BlueStalker replied:
in the existing functions, like listEnabledPrincipals, listEnabledTablePrivileges and other functions in this class, it already refers ConnectorIdentity in other purposes.

arhimondr commented on presto-hive-hadoop2/conf/files/hive-env.sh:
```diff
@@ -0,0 +1 @@
+export HADOOP_OPTS="$HADOOP_OPTS -Xmx1024m"
```
new line at the end

arhimondr commented on presto-hive-hadoop2/bin/run_hive_tests.sh:
```diff
@@ -7,6 +7,11 @@ set -euo pipefail -x
 cleanup_docker_containers
 start_docker_containers
 
+# restart HMS to pickup memory settings
```
Can this be added directly to the docker container to avoid this restart?

CC: @aweisberg

BlueStalker replied:
I think we need to directly bump the memory settings in the docker container

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/HiveTestUtils.java:
```diff
@@ -90,6 +90,8 @@ private HiveTestUtils()
     public static final ConnectorSession SESSION = new TestingConnectorSession(
             new HiveSessionProperties(new HiveClientConfig(), new OrcFileWriterConfig(), new ParquetFileWriterConfig()).getSessionProperties());
 
+    public static final String IMPERSONATIONUSER = "presto";
```
This shouldn't be needed once you refactor the statistics provider, remove it

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -738,95 +774,164 @@ public void dropRole(String role)
+    private MetastoreContext withContext(MetastoreContext context)
     {
-        return delegate.listTablePrivileges(databaseName, tableName, principal);
+        return isMetastoreImpersonationEnabled ? context : MetastoreContext.empty();
```
I'm not sure if it is correct to drop the entire context. How about we drop only the username if the impersonation is disabled?

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -738,95 +774,164 @@ public void dropRole(String role)
     }
 
-    public Set<HivePrivilegeInfo> loadTablePrivileges(String databaseName, String tableName, PrestoPrincipal principal)
+    private MetastoreContext withContext(MetastoreContext context)
```
Or better `private KeyAndContext<T> getCachingKey(MetastoreContext context, T key)`

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -114,20 +120,46 @@
+    {
+        for (HostAndPort key : metastoreStats.keySet()) {
+            String name = ObjectNames.builder(ThriftHiveMetastore.class)
+                    .withProperty("host", key.getHost())
```
Connector id should be included when exporting stats to avoid clashes when more than a sigle Hive catalog is deployed. See https://github.com/prestodb/presto/blob/master/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreModule.java#L71

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -114,20 +120,46 @@
     private final HiveCluster clientProvider;
     private final Function<Exception, Exception> exceptionMapper;
+    private final Map<HostAndPort, ThriftHiveMetastoreStats> metastoreStats;
+    private final boolean isHmsImpersonationEnabled;
```
s/isHmsImpersonationEnabled/impersonationEnabled/

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -114,20 +120,46 @@
+    private final MBeanExporter exporter;
     private final HiveCluster clientProvider;
     private final Function<Exception, Exception> exceptionMapper;
+    private final Map<HostAndPort, ThriftHiveMetastoreStats> metastoreStats;
```
metastoreHostStats

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -137,17 +169,21 @@ public ThriftHiveMetastoreStats getStats()
 
+    @Managed
+    @Flatten
+    public List<ThriftHiveMetastoreStats> getAllStats()
```
Remove this to avoid double export

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -158,17 +194,14 @@ public ThriftHiveMetastoreStats getStats()
-                        }
-                    }));
+                    .run("getDatabase", stats.getGetDatabase().wrap(() ->
+                            wrapMetastoreCallable(metastoreContext, (HiveMetastoreClient client, ThriftHiveMetastoreStats stat) -> Optional.of(stat.getGetDatabase().wrapCall(() -> client.getDatabase(databaseName))))));
```
It feels like we should find a better name for `wrapMetastoreCallable`. It is unclear what it does

BlueStalker replied:
getMetastoreClientThenCall

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -158,17 +194,14 @@ public ThriftHiveMetastoreStats getStats()
-                        }
-                    }));
+                    .run("getDatabase", stats.getGetDatabase().wrap(() ->
+                            wrapMetastoreCallable(metastoreContext, (HiveMetastoreClient client, ThriftHiveMetastoreStats stat) -> Optional.of(stat.getGetDatabase().wrapCall(() -> client.getDatabase(databaseName))))));
```
How about renaming the `wrapCall` to `execute`?

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -792,19 +809,26 @@ public void dropDatabase(String databaseName)
     }
 
+    @FunctionalInterface
+    public interface MetastoreCallable<V>
```
Let's put it next to the `wrapMetastoreCallable`

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -1000,19 +1020,33 @@ private void addPartitionsWithoutStatistics(String databaseName, String tableNam
+            HiveMetastoreClient client = clientProvider.createMetastoreClient(null);
+            return callable.call(client, metastoreStats.get(client.getAddress()));
+        }
+        try (HiveMetastoreClient client = clientProvider.createMetastoreClient(null)) {
```
Why do we need to keep this client open during the nested call?

How about

```
Optional<Token> delegationToken = Optiona.empty();

if(isHmsImpersonationEnabled){
  try (HiveMetastoreClient client = clientProvider.createMetastoreClient(null)) {
delegationToken = client.getMetastoreClient().get_delegation_token...
}
HiveMetastoreClient realClient = clientProvider.createMetastoreClient(token);
return callable.call(realClient, metastoreStats.get(realClient.getAddress()));

}


```

arhimondr replied:
Also let's move it all the way down, below all the usages

BlueStalker replied:
in your code, if impersonation disabled, it still call createMetastoreClient for twice, we should do it only once if it is disabled.  I don't think we need to keep the nested client open, is there any concerns?

BlueStalker commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/StaticHiveCluster.java:
```diff
@@ -62,16 +70,22 @@ public StaticHiveCluster(List<URI> metastoreUris, String metastoreUsername, Hive
     {
         List<HostAndPort> metastores = new ArrayList<>(addresses);
-        Collections.shuffle(metastores.subList(1, metastores.size()));
+        if (isMultipleMetastore) {
```
This is confusing code, the existing master code, is doing the shuffle from 1, (Collections.shuffle(metastores.subList(1, metastores.size()));), and then pick up the first and return.. I can remove the shuffle

BlueStalker commented on presto-hive-metastore/src/test/java/com/facebook/presto/hive/metastore/thrift/MockHiveMetastoreClient.java:
```diff
@@ -90,6 +92,18 @@ public int getAccessCount()
+    @Override
+    public Client getMetastoreClient()
+    {
+        return null;
```
function already removed.

BlueStalker commented:
@arhimondr Addressed most of the comments and please help another run, Also I don't find there is any new commits on docker-image repo, since the read timeout is still happening, I still believe we need to update the memory limit in the docker container for service like hms. @aweisberg do you have some updates on how to updating the docker-image?

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -1000,19 +1010,40 @@ private void addPartitionsWithoutStatistics(String databaseName, String tableNam
+            HiveMetastoreClient client = clientProvider.createMetastoreClient(Optional.empty());
+            return callable.call(client, metastoreHostStats.get(client.getAddress()));
+        }
+        try (HiveMetastoreClient client = clientProvider.createMetastoreClient(Optional.empty())) {
```
Why do we need to keep the `HiveMetastoreClient client` open for the entire period of the underlying call? Can it be closed right after retrieving the delegation token?

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreModule.java:
```diff
@@ -46,6 +47,7 @@ protected void setup(Binder binder)
         configBinder(binder).bindConfig(StaticMetastoreConfig.class);
 
         binder.bind(HiveMetastore.class).to(ThriftHiveMetastore.class).in(Scopes.SINGLETON);
+        binder.bindConstant().annotatedWith(ForThriftHiveMetastore.class).to(connectorId);
```
`ConnectorId` must already be bound. Let me know if that is not the case.

arhimondr commented on presto-hive/pom.xml:
```diff
@@ -412,6 +412,17 @@
+                <configuration>
+                    <!-- To allow usage of com.google.common.io.BaseEncoding -->
+                    <ignorePackages>
+                        <ignorePackage>com.facebook.presto.hive.authentication</ignorePackage>
```
Modernizer gives a legit suggesting that the https://docs.oracle.com/javase/8/docs/api/java/util/Base64.Encoder.html must be used instead of Guava.

BlueStalker replied:
Yes. that is the reason I put this line here in order to use Guava B64 Encoding. So what is the actual preference, use guava or use java.util.* ?

arhimondr replied:
When I suggested using Guava i didn't realize the Base64 encoder is included into Java standard library. Java standard library is preferred according to modernizr. In most cases we are trying to follow modernizr suggestions.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java:
```diff
@@ -16,26 +16,43 @@
 public class KerberosHiveMetastoreAuthentication
         implements HiveMetastoreAuthentication
 {
+    private static final Map<String, String> saslProperties = ImmutableMap.of(QOP, "auth", SERVER_AUTH, "true");
```
`SASL_PROPERTIES`

arhimondr reviewed (commented):
A couple of minor comments + missed comments from the previous review. Otherwise looks good to me.

BlueStalker commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -145,6 +152,7 @@ public static CachingHiveMetastore memoizeMetastore(ExtendedHiveMetastore delega
     private CachingHiveMetastore(
             ExtendedHiveMetastore delegate,
             ExecutorService executor,
+            boolean isMetastoreImpersonationEnabled,
```
What is the question here?

arhimondr replied:
Github must've done something weird. I was leaving this comment in the `memoizeMetastore` method

BlueStalker commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -738,95 +777,164 @@ public void dropRole(String role)
+    private MetastoreContext withContext(MetastoreContext context)
     {
-        return delegate.listTablePrivileges(databaseName, tableName, principal);
+        return isMetastoreImpersonationEnabled ? context : NO_IMPERSONATION_METASTORE_CONTEXT;
```
Hi, right now the caching key is including this MetastoreContext, and this method is just trying to use a single NO_IMPERSONATION_METASTORE_CONTEXT if the impersonation is not enabled. 

I do see the problems of including the whole MetastoreContext into the caching key since we are actually only using the identity (username) for it. But it is not a concern now for this PR. But we can change this class to include username only.

arhimondr replied:
Basically what I'm trying to suggest is to instead of trying to cache a single context in a static variable let's replace only the username if the impersonation is disabled. It is preffered to prevent accidental mistakes in the future if more fields is added to the context. So basically what I'm suggesting is something like this:


```
private KeyAndContext<T> getCachingKey(MetastoreContext context, T key) {
  String cachingUsername = isMetastoreImpersonationEnabled ? context.getUsername() : "no-impersonation-caching-user";
  MetastoreContext cachingContext = new MetastoreContext(cachingUsername);
  return new KeyAndContext(cachingContext, key);
}
```

BlueStalker commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -738,95 +774,164 @@ public void dropRole(String role)
     {
         try {
-            delegate.revokeTablePrivileges(databaseName, tableName, grantee, privileges);
+            delegate.revokeTablePrivileges(withContext(metastoreContext), databaseName, tableName, grantee, privileges);
```
yea, also removing all the unnecessary call of getCachingKey in this file

arhimondr commented on presto-hive-hadoop2/bin/run_hive_tests.sh:
```diff
@@ -7,6 +7,11 @@ set -euo pipefail -x
 start_docker_containers
 
+# restart HMS to pickup memory settings
+exec_in_hadoop_master_container cp /etc/hadoop/conf/hive-env.sh /etc/hive/conf/hive-env.sh
```
Ideally we should move docker related changes to the Docker definition itself. But let's not block this PR by that. Let's have a separate PR addressing that.

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/MetastoreClientConfig.java:
```diff
@@ -222,4 +224,17 @@ public MetastoreClientConfig setMetastoreCacheScope(MetastoreCacheScope metastor
+        return metastoreImpersonationEnabled;
+    }
+
+    @Config("hive.metastore.impersonation.enabled")
```
nit: `hive.metastore-impersonation-enabled` (same as other configs)

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -79,21 +81,24 @@
     private final boolean partitionVersioningEnabled;
+    private final boolean isMetastoreImpersonationEnabled;
+
+    private static final String NO_IMPERSONATION_USER = "no-impersonation-caching-user";
```
nit: let's move the static constant to the top (as we do for other static constants in other classes)

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -135,6 +143,7 @@ public static CachingHiveMetastore memoizeMetastore(ExtendedHiveMetastore delega
         return new CachingHiveMetastore(
                 delegate,
                 newDirectExecutorService(),
+                true,
```
This is incorrect. It must be set to the configuration value. While it currently doesn't effectively affect existing call sites it might at some point if it is used out of a "single query" context.

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -300,211 +311,215 @@ public void flushCache()
+    public Optional<Database> getDatabase(MetastoreContext metastoreContext, String databaseName)
     {
-        return get(databaseCache, databaseName);
+        return get(databaseCache, new KeyAndContext<>(getCachingKey(metastoreContext), databaseName));
```
Let's make the `getCachingKey` return the `KeyAndContext` that is used as a key for caching

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -300,211 +311,215 @@ public void flushCache()
     {
         try {
-            delegate.updateTableStatistics(databaseName, tableName, update);
+            delegate.updateTableStatistics(getCachingKey(metastoreContext), databaseName, tableName, update);
```
No need to wrap the context here

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -300,211 +311,215 @@ public void flushCache()
     {
         try {
-            delegate.updatePartitionStatistics(databaseName, tableName, partitionName, update);
+            delegate.updatePartitionStatistics(getCachingKey(metastoreContext), databaseName, tableName, partitionName, update);
```
No need to wrap the context here

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -300,211 +311,215 @@ public void flushCache()
     {
         try {
-            delegate.createDatabase(database);
+            metastoreContext = getCachingKey(metastoreContext);
```
No need to wrap the context here

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -300,211 +311,215 @@ public void flushCache()
     {
         try {
-            delegate.dropDatabase(databaseName);
+            metastoreContext = getCachingKey(metastoreContext);
```
No need to wrap the context here

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -300,211 +311,215 @@ public void flushCache()
     {
         try {
-            delegate.renameDatabase(databaseName, newDatabaseName);
+            metastoreContext = getCachingKey(metastoreContext);
```
No need to wrap the context here

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/StaticMetastoreConfig.java:
```diff
@@ -31,6 +31,7 @@
 
     private List<URI> metastoreUris;
     private String metastoreUsername;
+    private boolean isMultipleMetastoreEnabled;
```
Let's rename it accordingly, `metastoreLoadBalancingEnabled`, also getters and setters

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/StaticMetastoreConfig.java:
```diff
@@ -51,6 +52,19 @@ public StaticMetastoreConfig setMetastoreUris(String uris)
+        return isMultipleMetastoreEnabled;
+    }
+
+    @Config("hive.metastore.load-balancing.enabled")
```
nit: `hive.metastore.load-balancing-enabled`, also move this getters setters below the `getMetastoreUsername`, `setMetastoreUsername` to have the in the same order as the fields

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -137,17 +165,20 @@ public ThriftHiveMetastoreStats getStats()
     }
 
+    @VisibleForTesting
+    public List<ThriftHiveMetastoreStats> getAllStats()
```
Why not to return it as a map? It is very confusing what does this method return. How about returing a map, and calling the method `Map<HostAndPort, ThriftHiveMetastoreStats> getMetastoreHostStats`

arhimondr commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastoreClient.java:
```diff
@@ -52,15 +53,18 @@
 {
     private final TTransport transport;
     private final ThriftHiveMetastore.Client client;
+    private final HostAndPort address;
```
nit: Move this before the `transport` to maintain the same order of fields as in the constructor

arhimondr reviewed (commented):
LGTM % nits.

Please take a look at the test failures, they seem to be related

arhimondr reviewed (approved):
LGTM
